core/network-libp2p/src/custom_proto/handler.rs

```diff
@@ -165,44 +158,199 @@ struct PerProtocolNormalState<TMessage, TSubstream> {
 	shutdown: SmallVec<[RegisteredProtocolSubstream<TMessage, TSubstream>; 4]>,
 }
 
-impl<TMessage, TSubstream> PerProtocol<TMessage, TSubstream>
+impl<TMessage, TSubstream> PerProtocolNormalState<TMessage, TSubstream>
 where TMessage: CustomMessage, TSubstream: AsyncRead + AsyncWrite {
-	/// Enables the protocol. Returns an optional event to emit.
-	/// Must be passed the endpoint of the connection.
-	#[must_use]
-	fn enable(&mut self, endpoint: Endpoint)
-		-> Option<ProtocolsHandlerEvent<RegisteredProtocol<TMessage>, ProtocolId, CustomProtosHandlerOut<TMessage>>> {
+	/// Polls for things that are new. Same API constraints as `Future::poll()`.
+	/// Optionally returns the event to produce.
+	/// You must pass the `protocol_id` as we need have to inject it in the returned event.
+	/// API note: Ideally we wouldn't need to be passed a `ProtocolId`, and we would return a
+	/// 	different enum that doesn't contain any `protocol_id`, and the caller would inject
+	/// 	the ID itself, but that's a ton of code for not much gain.
+	fn poll(&mut self) -> Option<CustomProtoHandlerOut<TMessage>> {
+		for n in (0..self.pending_response.len()).rev() {
```

Comment:
Could be `while let Some(_) = self.pending_response.pop() {`? Any advantages of this approach that I'm missing?

Reply:
We're calling `self.pending_response.push(...)` within the loop. By doing it this way we only process each element once.